kitsune/wiki/utils.py

```diff
@@ -198,3 +200,51 @@ def get_visible_document_or_404(
 
 def get_visible_revision_or_404(user, **kwargs):
     return get_object_or_404(Revision.objects.visible(user, **kwargs))
+
+
+def build_topics_data(request, product, topics):
+    """Build topics_data for use in topic cards"""
+    topics_data = []
+
+    # Get all documents for all topics up front
+    all_docs, _ = documents_for(
+        request.user, request.LANGUAGE_CODE, topics=list(topics), products=[product]
+    )
+
+    # Convert all docs to Document objects
+    doc_ids = [doc["id"] for doc in all_docs]
+    documents = Document.objects.filter(id__in=doc_ids)
+
+    # Create a mapping of Document objects to their topics from all_docs
+    for doc_dict in all_docs:
+        for doc in documents:
+            if doc.id == doc_dict["id"]:
+                doc.topics_list = doc_dict.get("topics", [])
+                break
+
+    for topic in topics:
+        # Get featured articles first (returns up to 4 Document objects)
+        featured_articles = get_featured_articles(
+            product, locale=request.LANGUAGE_CODE, topic=topic
+        )
+
+        # Get docs for this topic from documents
+        topic_docs = [doc for doc in documents if topic.id in getattr(doc, "topics_list", [])]
+
+        # Only query for additional docs if we need more to reach 3
+        if len(featured_articles) < 3:
+            # If we don't have enough featured articles, append more from topic_docs
+            remaining_docs = [doc for doc in topic_docs if doc not in featured_articles]
```

Comment:
Nit. Since you are only going to use the first 3 documents of `featured_articles`, I would just add whatever remaining documents you need to get to 3.

Another way to do this might be to use `itertools.chain` like `chain(featured_articles, remaining_docs)[:3]` where you could make `remaining_docs` a generator instead of a list by wrapping it with `(...)` instead of `[...]`. That way you don't even need the `len(featured_articles) < 3` check. The `chain` only grabs documents from `remaining_docs` if needed.

Reply:
This is a great idea. Thanks.